accepted/00000-maintenance-windows.md

```diff
@@ -0,0 +1,99 @@
+- Feature Name: Maintenance Windows
+- Start Date: 2020-03-31
+- RFC PR: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Define Maintenance Windows in Uyuni and assign them to systems
+
+# Motivation
+[motivation]: #motivation
+
+Customers have company policies to perform changes on systems only at specific points in time.
+This is typically called "Maintenance Window".
+
+Uyuni should be able to define them and prevent executing of actions outside of such a window.
+
+# Architecture
+[architecture]: #architecture
+
+A system is in "Maintenance Mode" or not.
+When a system is in "Maintenance Mode" no action can be executed.
+Setting the Maintenance Mode "on" and "off" requires Organization Admin permissions.
+A "Maintenance Window" does nothing else than turning Maintenance Mode "on" when the window starts
+and "off" when it ends.
+
+Defining Maintenance Windows requires Organization Admin permissions.
+
+Maintenance Windows are defined on system level.
+Every System can get 1 Schedule assigned.
+A Schedule can be used for multiple systems.
+Inside of the Schedule, multiple Maintenance Windows can be defined.
+They can be defined as individual or as recurring events.
+
+To assign a Schedule to multiple systems, SSM should be used.
+
+
+# Detailed design
+[design]: #detailed-design
+
+## Implement a Maintenance Mode
+
+Check, how System Locking is implemented. Copy the mechanism for Maintenance Mode.
+Does it make sense to use System Locking as "Maintenance Mode"?
+
+For the first iteration we want to block all actions when the system is not in maintenance mode.
+In future iterations we can consider to allow some actions which are mostly read-only (package list refresh or hardware refresh).
+Also actions which prepare things for the maintenance window could be considered (pre-fetching packages for update).
+
+
+## Maintenance Schedule
+
+A Maintenance Schedule is a set of Maintenance Windows. A Maintenance Schedule has a name/label.
+The data should be stored as ICal data (https://tools.ietf.org/html/rfc5545).
+Importing data should be possible. (URL for external download?)
+
+We should use a java lib for ical like ical4j(https://github.com/ical4j/ical4j) to parse
+the ics files.
+Every `VEVENT` describe a maintenance window (or should we require a special `SUMMARY` value?).
+Events for a full day (anniversary) or multiple days should also be considered.
+Recurring events must be considered as well.
+
+We need to implement an overview page listing all schedules.
+We need the possibilities to add/edit/remove schedules.
+
+Watching details of a schedule could be done in a calendar widget.
+
+
+## Maintenance Window
+
+A Maintenance Window specify a time when changes on a system can be applied.
+It has a start date-time and an end date-time.
+It can be defined as recurring interval.
+
+When a Maintenance Window start, it turn the maintenance mode "on".
```

Comment:
We also need to clarify "how" does the switch happen. We need some mechanism, that is "triggered" when the maint. window starts/ends. The first idea that comes to my mind is that when we create the maintenance window, we need to create 2 taskomatic schedules (1 for turning on the maint. mode for the systems, 1 for turning it off). The one problem I could foresee now is that taskomatic only support cron expressions and I fear iCal allows defining more elaborate frequencies (last day of month).

Reply:
Do we really need a "value" or could it be a function `isInMaintenanceMode()` which lookup the calendar to decide if it is or not?

Reply:
Agreed we could perform this check every time when scheduling an action and would not need this `maintenance mode` flag. The org admin could get a warning in this case (could decide to perform the action anyway), other users would get an error stating that the system isn't in `maintenance mode` at that time.

Reply:
From a usage point of view, there are two different interactions I can imagine:

- admin defines a maintenance schedule and expects the system to be in maintenance mode at the given times
- admin (or an API call triggered by a third party tool/scheduler) puts the system into planned or emergency maintenance mode _now_ or switches it off again.

Of course you can implement both in different ways. For example, if there's only a method to check whether the system has a maintenance mode every time an action is triggered, you'll need to create a (temporary) schedule when the system is set to maintenance manually.

But it may be easier to do it the other way around: Manual maintenance mode changes the flag immediately, and the schedules are checked at certain intervals to set/unset maintenance mode as needed.

You may want to look into the potential performance implications of both, for a typical scenario (i.e. is it going to create more load to set/unset the flags for thousands of machines when a window is hit or to calculate the status on the fly based on the schedule). There must be prior art. Calendaring tools are doing this (busy status) all the time. 😄 

Additional considerations:

- Would we want to bring back a system from maintenance mode only at the end of the window or should there be away of switching it off once all maintenance tasks have been finished (which could be much earlier than the end of the window)? I'd think that most admins would _love_ to have the systems come back sooner (which could also be used in uptime statistics to prove that SLAs are met/improving). But the corner case is when an admin figures out later (but still within the maintenance window originally scheduled) that there's more to do.
- What happens if a system is in manual maintenance mode and the scheduled maintenance window kicks in? I guess an admin wouldn't want a system to come back from maintenance automatically at the end of the window if the reason for the manual override hasn't been resolved yet.
- What if a long-running task _exceeds_ the scheduled window and can't be rolled back quickly? In other words, unless the task gets finished, the system will be in an unusable state. Yes, this should never happen. But if it _does_, I'd think the admin would rather see the system flagged as in maintenance until things are finished than get the system back to production while it's still broken.

In the end, this could mean that we need to separate between "actual maintenance mode" and "scheduled maintenance mode" (with the actual one being driven manually or through exit criteria (e.g. full maintenance task action chain has been completed successfully) and the scheduled on just indicating what _should_ be based on the schedule.

Reply:
To be able to plan, we need to know about maintenance windows upfront.
A 3rd party application which just set the maintenance mode "on" and "off"
is not helpful.

The emergency case is special. It can set the maintenance mode "now".
You cannot predict when an emergency happens, so there also no need to "plan"
actions for it.
When it happen, you set the system in maintenance mode and create new actions
wich should be exectuted "now".

This means when we check the schedule for maintenance windows we need to consider
the current state. If the system is currently in maintenance mode, we allow to schedule
actions "now" - maybe also for the next minute.

I will add a section about this case